### PR TITLE
VM: Add BaseStateManager

### PR DESCRIPTION
**Pull request by holgerd77**
Part of #1533 

This PR is adding a `BaseStateManager` class as outlined in the diagram I posted on the Discord #stateless channel extracting out non-backend related functionality from the `DefaultStateManager`. This will make it easier for us to exchange the state manager backend e.g. with a verkle trie and will generally make it easier to do custom state manager implementations (my side hope is that we can then really easily implement an RPC state manger which we might also want to directly add to the code base).

WIP

### CODE REVIEW

**Comment by holgerd77** [comment]
Ok, rebased this and did some cleanup. This is now open for an early review. 😄 

Please do not merge yet since we (likely) don't want to have this yet in the imminent ArrowGlacier HF (and other stuff) releases, makes more sense to work more with this along the Verkle Trie work and see if/how this fits.

One bigger last change I made: I wondered if I could get the `generateGenesis()` function to the `BaseStateManager` as well - see bcbf341 and subsequent commit - but when I first changed this to direct `putAccount()` calls interestingly enough the following [eth_getBalance](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/client/test/rpc/eth/getBalance.spec.ts#L40) test in the client started to fail.

I digged a bit deeper. This was due to the `putAccount()` method only writing to the StateManager cache. The VM in the [eth_getBalance implementation](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/client/lib/rpc/modules/eth.ts#L346) is copied over, but along this copy the `Cache` from the `StateManager` is not taken over with the StateManager (see also comment on the code docs of the [StateManager.copy()](https://github.com/ethereumjs/ethereumjs-monorepo/blob/master/packages/vm/src/state/stateManager.ts#L107) function). So only state up to the latest commit is taken over.

I now solved this by doing a `cache.flush()` call at the end of the `generateGenesis()` function, I would say this is a clean solution.

As some side thought we can generally question if the cache shouldn't also be copied over on the `StateManager.copy()` function? 🤔 Maybe this is generally causing some problems here and there? I would assume that no one is using this with this limitation in mind (neither we ourselves nor - even more unlikely - community users).

**Comment by holgerd77** [comment]
Will also circle in @fvictoro, this might generally be interesting for you guys to ease your custom StateManager implementation.

**Comment by holgerd77** [comment]
Will also circle in @fvictorio, this might generally be interesting for you guys to ease your custom StateManager implementation.

**Comment by ryanio** [comment]
> As some side thought we can generally question if the cache shouldn't also be copied over on the `StateManager.copy()` function? 🤔 Maybe this is generally causing some problems here and there? I would assume that no one is using this with this limitation in mind (neither we ourselves nor - even more unlikely - community users).

Ah nice observation, haven't thought about this.

When I have been using `vm.copy()` (e.g. in the client) I am generally using it from a clean state point of view or using `setStateRoot` to set the state to specifically where I want it.

I'm not sure the use case where I'd also want the existing cache to be available, but if we find a use for it we could introduce it as a boolean option (with default to current behavior). I'll keep it in mind if/when I start to work on a "pending block" feature in the client if it could be useful there.

**Comment by ryanio** [comment]
reviewed the code and lgtm! I'll hold off on an approval until we hear back from hardhat if there is any further feedback or requested changes.

**Comment by holgerd77** [comment]
> I'm not sure the use case where I'd also want the existing cache to be available, but if we find a use for it we could introduce it as a boolean option (with default to current behavior). I'll keep it in mind if/when I start to work on a "pending block" feature in the client if it could be useful there.

I would rather think this doesn't even need a flag and can eventually be seen as some wider bug/behavior fix, since this doesn't take anything away from the existing behavior - correct me if I am wrong - but only adds on being more flexible (respectively also: less error prone) on a `copy()` operation.

File: packages/vm/src/state/baseStateManager.ts
Comment (acolytec3) [inline packages/vm/src/state/baseStateManager.ts:104]: This is nitpicky but why are we not just using `true/false` here as returned by the `isContract` and `isEmpty` methods instead of "yes/no"?  This is purely stylistic but true/false just feels like a more helpful reporting of the state of this account address than saying "yes/no".

**Review by acolytec3 — APPROVED** [review]
Looks good to me (with one nitpicky comment about debug messaging that you can take or leave). For a sanity check,  I ran this branch for a few thousand blocks on mainnet client and everything seemed fine.

File: packages/vm/src/state/baseStateManager.ts
Comment (holgerd77) [reply packages/vm/src/state/baseStateManager.ts:104]: I found it a little bit more readable to have `yes/no` here (but brain needs to do a tiny bit less of internal mapping in this occasion 😛), but I have absolutely "no" opinion on things like that, happy with the one way or the other.

Will leave this now to not dismiss the review.

**Comment by holgerd77** [comment]
Ok, I will now merge this in so that we can build upon this.

**Pull request merged**
